feat(texteditor): Markdown <-> JSON dual conversion for backward compatibility

habibayman opened the pull request:
## Summary
This PR includes the conversion of the structured JSON format TipTap uses to&from Markdown to maintain backward compatibility with the old editor's API.

supports the conversion for:
- Standard Markdown elements previously handled by the ToastUI editor and its Showdown converter.
- A specific, legacy format for custom nodes, particularly for Images `(![alt](placeholder/checksum.ext =WxH))` and Math Formulas `$$latex$$`

The formats for the custom nodes are adapted from the old editor's standard syntax conversion.

…

## References
This PR adds no changes to the UI that will remain permanently, just a playground to test in!
- part of the tracking issue: #5049 
- fixes: #5185 

…

## Reviewer guidance

…

nucleogenesis reviewed (commented):
I've tested things out locally w/ images and various markdown updates. I did some testing around how things work when there is a lot of text. 

My not very rigorous test suggests that things work well up to a pretty significant document size - but once I'd copy-pasted the default text in there a few dozen times it started to get a bit laggy - which I think is to be expected at that point :sweat_smile: 

I added a couple of notes on code review and will give a little more time to the markdown serialization, but all-in-all this feels and looks great!

habibayman commented on contentcuration/contentcuration/frontend/shared/views/TipTapEditor/TipTapEditor/components/image/ImageNodeView.vue:
```diff
@@ -145,10 +205,12 @@
 
         clearTimeout(debounceTimer);
-        debounceTimer = setTimeout(saveWidth, 500);
+        debounceTimer = setTimeout(saveSize, 500);
```
I use this so I don't clatter the undo/redo stack. If I keep spamming it with updates, using the undo button after that will be a nightmare!

The `saveSize()` function causes an editor state update which by default gets added to a stack.

habibayman commented on pnpm-lock.yaml:
```diff
@@ -5026,6 +5029,11 @@ packages:
 
+  marked@16.1.1:
+    resolution: {integrity: sha512-ij/2lXfCRT71L6u0M29tJPhP0bM5shLL3u5BePhFwPELj2blMJ6GDtD7PfJhRLhJ/c2UwrK17ySVcDzy2YHjHQ==}
+    engines: {node: '>= 20'}
```
Can you run the linter and see if you can replicate having similar lines in the console?

```
Error while parsing /home/habiba/Downloads/studio/node_modules/marked/lib/marked.esm.js
Line 247, column 12: Unexpected token ;
`parseForESLint` from parser `/home/habiba/Downloads/studio/node_modules/.pnpm/vue-eslint-parser@9.4.3_eslint@8.57.1/node_modules/vue-eslint-parser/index.js` is invalid and will just be ignored
```

My first guess was that this line must be the problem; so I tried downgrading marked to a version that uses `node > 18` which should work, but that changed nothing.

habibayman commented:
@nucleogenesis 

> which I think is to be expected at that point 😅

This may be the cost of all these regular expressions? I think it's tolerable though. 

I added unit tests for loading/saving markdown, and added a test for a large input to see whether it will tolerate that or not 👍🏻

nucleogenesis reviewed (commented):
@habibayman Overall this is looking great - the additional tests are solid and cover the serialization well, thanks for adding those. 

I think that this is good to go - I'll hold off on the final Approval until great mergening is upon us.

marcellamaki reviewed (approved):
Officially approving now after merging to studio has been unblocked
